docs/proposals/20230123-azure-service-operator.md

```diff
@@ -0,0 +1,252 @@
+        - [Reconciliation](#reconciliation)
+        - [API Changes](#api-changes)
+        - [Installation](#installation)
+        - [Caveats](#caveats)
+    - [Security Model](#security-model)
+    - [Risks and Mitigations](#risks-and-mitigations)
+  - [Alternatives](#alternatives)
+      - [Adopt new SDK directly](#adopt-new-sdk-directly)
+      - [Crossplane](#crossplane)
+  - [Upgrade Strategy](#upgrade-strategy)
+  - [Additional Details](#additional-details)
+    - [Test Plan](#test-plan)
+    - [Graduation Criteria](#graduation-criteria)
+  - [Implementation History](#implementation-history)
+
+## Glossary
+
+- **Azure Service Operator (ASO)** - ([link](https://azure.github.io/azure-service-operator/)) a collection of Kubernetes CRDs and controllers which extend the Kubernetes API to manage Azure resources like virtual machines and resource groups.
+- **Bring-your-own (BYO) resource** - an Azure resource backing a Cluster API cluster but managed by the user. CAPI and CAPZ should not modify or delete this resource.
+
+## Summary
+
+This proposal describes how CAPZ could utilize Azure Service Operator (ASO) to create, read, update, and delete Azure resources with the Kubernetes API and custom resources defined and reconciled by ASO.
+
+## Motivation
+
+As a Cluster API infrastructure provider, CAPZ's functionality can be broadly divided into:
+1. Translating Cluster API resource definitions into Azure infrastructure definitions
+2. Interfacing with the Azure platform to manage creating, updating, and deleting that infrastructure
+
+Currently, CAPZ handles `2.` with Azure API calls, which requires ensuring dependencies between Azure resources are met and maintaining state to asynchronously handle long-running operations. Because ASO's primary objective is to handle `2.`, CAPZ could then more declaratively specify all of the infrastructure needed for a given Cluster API resource and let ASO handle the finer details of ensuring that desired state is achieved in Azure. By then primarily focusing on `1.`, CAPZ reduces its overall code maintenance burden and affirms ASO as the canonical way to manage Azure infrastructure from Kubernetes.
+
+In addition, the [Azure Go SDK](https://github.com/Azure/azure-sdk-for-go) has recently deprecated its set of APIs that CAPZ currently uses in favor of a new set of backwards-incompatible APIs. Because the deprecated Azure Go SDK APIs are no longer receiving updates to adopt new Azure API versions, CAPZ must adopt the new APIs to be able to expose the newest features in Azure to its users. Because ASO already uses the new APIs, adopting ASO in CAPZ accomplishes the goal of migrating off of the deprecated APIs.
+
+### Goals
+
+- Logically separate transformation of Cluster API resources to infrastructure definitions from reconciliation of that infrastructure in Azure
+- Determine a method to migrate CAPZ's individual Azure service interfaces one-by-one to use ASO
```

Comment:
Maybe we'll discover that "one-by-one" is not strictly practical, and in fact the project will be better served by doing some of the iteration in batches. So perhaps the more accurate way to state this would be "Determine a method to migrate CAPZ's individual Azure service interfaces to use ASO iteratively, and not all at once as a major code change that affects all Azure service code".